openedx/core/djangoapps/session_inactivity_timeout/middleware.py

```diff
@@ -34,27 +39,67 @@ def process_request(self, request):
+            #   A user could be inactive for almost SESSION_ACTIVITY_SAVE_DELAY_SECONDS but since their session 
+            #   isn't being saved during that time, their last activity timestamp isn't being updated.
+            #   When they hit the inactivity timeout, it will be based on the last saved activity time.
+            #   So the effective timeout could be as short as: SESSION_INACTIVITY_TIMEOUT_IN_SECONDS - SESSION_ACTIVITY_SAVE_DELAY_SECONDS.
+            #   This means users might be logged out earlier than expected in some edge cases.
+            # .. setting_warning:  Must be smaller than SESSION_INACTIVITY_TIMEOUT_IN_SECONDS.
+            frequency_time_in_seconds = getattr(settings, "SESSION_ACTIVITY_SAVE_DELAY_SECONDS", 900)
             # what time is it now?
             utc_now = datetime.utcnow()
 
             # Get the last time user made a request to server, which is stored in session data
-            last_touch = request.session.get(LAST_TOUCH_KEYNAME)
+            last_touch_str = request.session.get(LAST_TOUCH_KEYNAME)
 
             # have we stored a 'last visited' in session? NOTE: first time access after login
             # this key will not be present in the session data
-            if last_touch:
-                # compute the delta since last time user came to the server
-                time_since_last_activity = utc_now - last_touch
-
-                # did we exceed the timeout limit?
-                if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
-                    # yes? Then log the user out
-                    del request.session[LAST_TOUCH_KEYNAME]
-                    auth.logout(request)
-                    return
-
-            request.session[LAST_TOUCH_KEYNAME] = utc_now
+            if last_touch_str:
+                try:
+                    last_touch = datetime.fromisoformat(last_touch_str)
+                    time_since_last_activity = utc_now - last_touch
+
+                    # did we exceed the timeout limit?
+                    has_exceeded_timeout_limit = time_since_last_activity > timedelta(seconds=timeout_in_seconds)
+                    if has_exceeded_timeout_limit:
+                        del request.session[LAST_TOUCH_KEYNAME]
+                        auth.logout(request)
+                        return
+                except (ValueError, TypeError) as e:
+                    log.warning("Parsing last touch time failed: %s", e)
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
+            else:
+                monitoring_utils.set_custom_attribute('session_inactivity.first_login', True)
+                log.debug("No previous activity timestamp found (first login)")
+
+            # Store activity timestamp
+            request.session[LAST_TOUCH_KEYNAME] = utc_now.isoformat()
+
+            # Periodically allow a full save (every n seconds)
```

Comment:
1. This comment doesn't feel appropriate to these particular lines.
2. I'd replace comments with clear variable names like `has_exceeded_timeout_limit` above.
  - Consider `proceed_with_period_save` as one variable.
  - Maybe `has_save_delay_been_exceeded` (or something like that, could be an additional variable?